files/audit.sh

```diff
@@ -0,0 +1,178 @@
+                       -e will be a noop
+
+  [-v|--verbose]       Print more info
+
+  [-h|--help]          Print this help message
+ENDHELP
+exit
+}
+
+croak() {
+    echo "ERROR - $*"
+    exit 99
+}
+
+# Usage:
+# $1 : String to append to email subject line (brief status info)
+# All other arguments past 1st : Arguments to print in email body, each string argument will
+# get its own line, can pass "" to print an empty line
+email_and_exit() {
+    subject_status="${1}"
+    shift
+    email_message=("$@") # Feed all other arguments into array
+
+    if [[ $verbose ]]; then
+        echo -e "${subject_status} \n $(for i in "${email_message[@]}";do echo "$i"; done)"
+
+        if lsdef  -t site -i excludenodes | grep -q excludenodes; then
+            email_message=("${email_message[@]}" "" "Nodes excluded because they are in excludenodes set in site table:" "$(lsdef  -t site -i excludenodes | grep excludenodes)")
+        fi
+
+        if [[ ${#NODES_TO_IGNORE[@]} -ne 0 ]]; then
+            email_message=("${email_message[@]}" "Nodes ignored by being set in NODES_TO_IGNORE:" "${NODES_TO_IGNORE[@]}")
+        fi
+    fi
+
+cat << EOF | mail -s "${MAIL_SUBJECT} - ${subject_status}" "${MAIL_TO}"
+$(for i in "${email_message[@]}";do echo "$i"; done)
+EOF
+
+exit
+}
+
+# Get CLI options
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        -a|--automated) automated=0 ;;
+        -m|--module_audit) module_audit=0 ;;
+        -e|--email_compare) email_compare=0 ;;
+        -v|--verbose) verbose=0 ;;
+        -h|--help) show_help ;;
+        *) echo "Unknown parameter passed: $1"; show_help;;
+    esac
+    shift
+done
+
+# Sanity checks
+[[ $EUID -ne 0 ]] && croak "This script should be ran as root via sudo"
+[[ -z "${SUDO_USER}" && ! $automated ]] && croak "Could not determine SUDO_USER, this script should be ran via sudo"
+[[ -z "$AUDIT_SCRIPT" ]] && croak "AUDIT_SCRIPT not defined in config file"
+[[ $email_compare ]] && [[ -z "$MAIL_TO" ]] && croak "Email requested but MAIL_TO not defined in config file"
+if [[ $module_audit ]]; then
+    [[ -z "$MODULE_HOST" ]] && croak "Module audit requested but MODULE_HOST is not defined in config file"
+    [[ -z "$MODULE_DIRS" ]] && croak "Module audit requested but MODULE_DIRS is not defined in config file"
+fi
+
+
+mapfile -t NODES_IN_ALL_GRP < <(/opt/xcat/bin/nodels all)
+
+if [[ $automated ]]; then
+    AUDIT_USER="root"
+    REPORT_DST="$AUTOMATED_REPORT_DST/$DATETIME"
+else
+    AUDIT_USER="${SUDO_USER}"
+    REPORT_DST="$BASE_REPORT_DST/$DATETIME-$AUDIT_USER"
+fi
+
+
+if [ ! -d "$REPORT_DST" ]
+then
+   echo "Saving reports to $REPORT_DST"
+   mkdir -p "$REPORT_DST"
+fi
+
+# Build list of nodes to perform audit on
+for ignore_node in "${NODES_TO_IGNORE[@]}"; do
+    for i in "${!NODES_IN_ALL_GRP[@]}"; do
+        if [[ ${NODES_IN_ALL_GRP[$i]} == "$ignore_node" ]]; then
+            unset 'NODES_IN_ALL_GRP[i]'
+        fi
+    done
+done
+NODES=("${NODES_IN_ALL_GRP[@]}")
```

Comment:
This seems a bit inefficient, as does requiring that nodes in cfg_nodes_to_ignore be listed individually.

An alternative might be allowing cfg_nodes_to_ignore to be an array containing either nodes or node ranges (w/o commas, which might be hard to enforce). E.g.:
```
profile_xcat::master::inventory_audit::cfg_nodes_to_ignore:
  - node[01-02]
  - node[46-53]
```

Then I guess loop through that list and construct a node list like `all,-node[01-02],-node[46-53]`. I.e., leverage xCAT's `nodels` function a bit more.

I think those types of node ranges are readable enough.

On the other hand I may be missing advantages to listing the ignore nodes all singly. And I'm not at all suggesting it's actually worthwhile to code such a change. Just sharing an idea.

Reply:
Nice I didn't know about the `all,-node[01-02],-node[46-53]` syntax, I'll take a look

Reply:
Ok now using `nodels` to do more of the work with removing nodes. Supports individual nodes, node ranges, or xcat groups.

Also added the ability to list the nodes you want to audit in the same way, so in hiera they could be listed as xcat groups, list of individual nodes, or list of node ranges (or any combination of those)